File: src/types/_ActivatableSkillTargetCategory.ts
import * as DB from "tsondb/schema/dsl"
import { NestedTranslationMap } from "./Locale.js"
import { TargetCategoryIdentifier } from "./_Identifier.js"

export const AffectedTargetCategories = DB.TypeAlias(import.meta.url, {
  name: "AffectedTargetCategories",
  comment: `The target category – the kind of creature or object – the skill affects.

If no target categories are given, the skill applies to all target categories.`,
  type: () => DB.Array(DB.IncludeIdentifier(SpecificAffectedTargetCategory)),
})

const SpecificAffectedTargetCategory = DB.TypeAlias(import.meta.url, {
  name: "SpecificAffectedTargetCategory",
  type: () =>
    DB.Object({
      id: DB.Required({ type: DB.IncludeIdentifier(SpecificAffectedTargetCategoryIdentifier) }),
      translations: NestedTranslationMap(
        DB.Optional,
        "SpecificAffectedTargetCategory",
        DB.Object({
          note: DB.Required({ type: DB.String({ minLength: 1 }) }),
        }),
      ),
    }),
})

const SpecificAffectedTargetCategoryIdentifier = DB.Enum(import.meta.url, {
  name: "SpecificAffectedTargetCategoryIdentifier",
  values: () => ({
    Self: DB.EnumCase({ type: null }),
    Zone: DB.EnumCase({ type: null }),
    LiturgicalChantsAndCeremonies: DB.EnumCase({ type: null }),
    Cantrips: DB.EnumCase({ type: null }),
    Spellworks: DB.EnumCase({ type: null }),
    Predefined: DB.EnumCase({ type: TargetCategoryIdentifier() }),
  }),
})

File: Sources/OptolithDatabaseSchema/Types/_ActivatableSkillTargetCategory.swift
import FileDB

/// The target category – the kind of creature or object – the skill affects.
///
/// If no target categories are given, the skill applies to all target categories.
typealias AffectedTargetCategories = [SpecificAffectedTargetCategory]

@Embedded
public struct SpecificAffectedTargetCategory {
    let id: SpecificAffectedTargetCategoryIdentifier
    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]?

    struct Translation {  // SpecificAffectedTargetCategoryTranslation
        @MinLength(1)
        let note: String
    }
}

@ModelEnum
public enum SpecificAffectedTargetCategoryIdentifier {
    case `self`
    case zone
    case liturgicalChantsAndCeremonies
    case cantrips
    case predefined(TargetCategoryIdentifierObject)
}
